contrib/babelfishpg_common/src/spatialtypes.c

```diff
@@ -67,134 +113,1465 @@ load_functions()
+    }
+
+    /* Extract SRID (little-endian) */
+    srid = (input_data[3] << 24) | (input_data[2] << 16) | 
+           (input_data[1] << 8) | input_data[0];
+
+    /* Extract and convert latitude */
+    memcpy(&lat_bits, input_data + 6, sizeof(uint64_t));
+    lat_bits = le64toh(lat_bits);
+    memcpy(&lat, &lat_bits, sizeof(double));
+
+    while (byte_position < input_len)
+    {
+        memcpy(input_coord, input_data + byte_position, 8);
+        if (memcmp(input_coord, empty_geom, 8) == 0)
+        {
+            isNaN = 1;
+            break;
+        }
+        byte_position += 8;
+    }
+
+    /* Extract and validate geography type */
+    memcpy(geom_type, input_data + 4, 2);
+
+    /* Determine dimension flag based on geography type */
+    if (memcmp(geom_type, "\x01\x04", 2) == 0) {
+        dimension_flag = 0;      /* XY */
+    } else if (memcmp(geom_type, "\x01\x0c", 2) == 0) {
+        dimension_flag = 1;      /* XY (alternate) */
+    } else if (memcmp(geom_type, "\x01\x0d", 2) == 0) {
+        dimension_flag = 2;      /* XYZ */
+    } else if (memcmp(geom_type, "\x01\x0e", 2) == 0) {
+        dimension_flag = 3;      /* XYM */
+    } else if (memcmp(geom_type, "\x01\x0f", 2) == 0) {
+        dimension_flag = 4;      /* XYZM */
+    } else {
+        ereport(ERROR,
+                (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+                 errmsg("Unsupported geography type")));
+    }
+
+    /* Validate SRID against allowed values */
+    valid_srids_array = DatumGetArrayTypeP(DirectFunctionCall1(get_valid_srids, 
+                                           Int32GetDatum(0)));
+    valid_srids = (int32 *)ARR_DATA_PTR(valid_srids_array);
+    num_valid_srids = ArrayGetNItems(ARR_NDIM(valid_srids_array), 
+                                     ARR_DIMS(valid_srids_array));
+
+    for (int i = 0; i < num_valid_srids; i++) {
+        if (valid_srids[i] == srid) {
+            srid_valid = true;
+            break;
+        }
+    }
+
+    /* Process geography if SRID is valid and coordinates are valid */
+    if (srid_valid && !isNaN && lat >= -90.0 && lat <= 90.0) {
+        if (dimension_flag == 0) {
+            /* Handle empty geography */
+            result = (bytea *) palloc(VARHDRSZ + 25);
+            SET_VARSIZE(result, VARHDRSZ + 25);
+            result_data = (uint8 *)VARDATA(result);
+
+            /* Construct new geography with updated header */
+            memcpy(result_data, new_header, 5);
+            memcpy(result_data + 5, input_data, 4);
+            memcpy(result_data + 9, empty_geom, 8);
+            memcpy(result_data + 17, empty_geom, 8);
+        } else {
+            /* Handle non-empty geography with dimensions */
+            
+            /* Adjust header based on dimension type */
+            if (dimension_flag == 2) new_header[4] = 0xA0;  /* XYZ */
+            if (dimension_flag == 3) new_header[4] = 0x60;  /* XYM */
+            if (dimension_flag == 4) new_header[4] = 0xE0;  /* XYZM */
```

Comment:
Consider using an array based variable for different headers and assign it directly using `dimension_flag` as index to remove redundant if checks.